### PR TITLE
feat: update how attentionEl's position should be controlled

### PR DESCRIPTION
**Pull request by felicia-haggqvist**
**Part of fixes for Jira-tickets:** [WARP-432](https://nmp-jira.atlassian.net/browse/WARP-432), [WARP-341](https://nmp-jira.atlassian.net/browse/WARP-341) & [WARP-373](https://nmp-jira.atlassian.net/browse/WARP-373)

- Update `@floating-ui/dom` to latest version (`1.6.3`)
- Add more `placement`-options for Attention: `top-start`, `top-end`, `right-start`, `right-end`, `left-start`, `left-end`, `bottom-start`, `bottom-end`
- Update type for `Direction` to also include these new placement-options
- Add `distance`, `skidding`, `flip`, `fallbackPlacements` to the attentionState type (Note: these are new props from the AttentionEl and are used in `floating-ui` to give more flexibility for the user when positioning the attentionEl.) 
- Move in the calculation of `arrowRotation` and `arrowDirectionClassname` to the `warp-ds/core` repo instead of having duplicated logic in `warp-ds/react` `warp-ds/vue`, `warp-ds/elements`
- Add more options in the flip-middleware that is part of the `computePosition` (in the `useRecompute` function) so that user can choose whether the attentionEl should flip its placement and if so, which preferred placement should it flip to ( See more detail on how you can configure the middleware in [floating-ui's documentation).](https://floating-ui.com/docs/middleware)
- Remove `transform: `translate3d(${Math.round(position.x)}px, ${Math.round(position.y)}px, 0)`` that was causing the position of the attentionEl to not be aligned correctly. 
- Refactor the calculation of the arrowEl's position to be able to account for when `placement` includes `-start` or `-end` and that it is aligned correctly when the attentionEl's placement is flipped
- Add `autoUpdatePosition` function that uses `floating-ui`'s `autoUpdate` function (See more info on this function [here](https://floating-ui.com/docs/autoUpdate)) that makes it possible to recalculate the attentionEl's placement when resizing the window or on scroll. 

**To test:**
- Link with either `warp-ds/react` (branch: `feat/control-attention-position`), `warp-ds/vue` (branch: `feat/conrol-attention-position`)or `warp-ds/elements` (branch: `feat/control-attention-position`)
- Make sure that `warp-ds/react`, `warp-ds/vue` and `warp-ds/elements` are also linked to `warp-ds/css` (branch: `feat/control-attention-position`).

### CODE REVIEW

File: package.json
Comment (BalbinaK) [inline package.json]: Due to problems with renovate and pnpm, we should probably pin this dependency (remove `^`). Also, I see that `@floating-ui/dom` is now at 1.6.3, so we could then pin it to that version (https://www.npmjs.com/package/@floating-ui/dom).

File: src/attention/utils/helpers.ts
Comment (BalbinaK) [inline src/attention/utils/helpers.ts]: It's a small thing, but maybe we could add underscore as separator, like `TOP_START`? I think it would be easier to read.

File: src/attention/utils/helpers.ts
Comment (BalbinaK) [inline src/attention/utils/helpers.ts]: was a type required here? `any` doesn't seem to help much 😄

File: src/attention/utils/helpers.ts
Comment (BalbinaK) [inline src/attention/utils/helpers.ts]: It doesn't seem like these need to be defined in `AttentionState` type. At least I couldn't see them being used in any of the utils.

File: src/attention/utils/helpers.ts
Comment (BalbinaK) [inline src/attention/utils/helpers.ts]: Would it make sense to type `name` as `Directions`?

File: src/attention/utils/helpers.ts
Comment (BalbinaK) [inline src/attention/utils/helpers.ts]: Do we need to write `: ReferenceElement` if it's already defined with `as ReferenceElement`?

File: src/attention/utils/helpers.ts
Comment (BalbinaK) [inline src/attention/utils/helpers.ts]: Is it valid to type something `as unknown as HTMLElement`?

**Review by BalbinaK — COMMENTED** [review]
Still reviewing, but thought I'd comment on the first half before I'm done 😊

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: I added type `any` to rotation because it was complaining on line 130, when I'm setting `arrowRotation` to `rotation[arrowDirection]`: 
```
"Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ "left-start": number; left: number; "left-end": number; "top-start": number; top: number; "top-end": number; "right-start": number; right: number; "right-end": number; "bottom-start": number; bottom: number; "bottom-end": number; }'.
  No index signature with a parameter of type 'string' was found on type '{ "left-start": number; left: number; "left-end": number; "top-start": number; top: number; "top-end": number; "right-start": number; right: number; "right-end": number; "bottom-start": number; bottom: number; "bottom-end": number; }'."
```

I wasn't sure how to type it otherwise in order to remove the error. Do you have any suggestions for a better type? :)

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: It was complaining if I removed the `as ReferenceElement` since targetEl has type unknown in `AttentionState` because it can also be undefined. If I remove `as ReferenceElement`, I get this error: `"Type 'unknown' is not assignable to type 'ReferenceElement'."`

File: src/attention/utils/helpers.ts
Comment (BalbinaK) [reply src/attention/utils/helpers.ts]: Maybe something like `Record<Directions, number>`?

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: I think you can assert it to first `unknown`, but it is considered an unsafe type assertion, so I'm overriding TypeScript by first asserting the variable as unknown and then to the desired type, which in this case is `HTMLElement.` However, I just found [this article](https://medium.com/@InspireTech/exploring-type-annotation-and-type-assertion-in-typescript-a-comprehensive-comparison-4d56d2b11071#:~:text=Type%20Assertion%3A%20(Not%20recommended),from%20one%20type%20to%20another.) and saw that is not recommended to use type assertion. It is instead recommended to use type annotation. So I think I will change it to this: 
```
const floatingEl = state?.attentionEl
const arrowEl = state?.arrowEl
```

And then I will just have a check for `  if (!state?.targetEl || !state?.attentionEl) return` before I define the `referenceEl`, `floatingEl` and `arrowEl`

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts:145]: I am not sure how to remove the type assertion for `targetEl`, since that one has type `unknown` in, for example, our react-repo, and if I need to set it already in the react-repo to type `ReferenceElement` I would need to add `floating-ui/dom` devDependency to that project just for that. Type `ReferenceElement` is defined in `floating-ui/dom`. 

A potential solution could be to set the type of `targetEl` to `ReferenceElement` here in the core-repo that is part of the `AttentionState` type, and in the react-repo I could set the type of `targetEl` to what `floating-ui/dom` has defined `ReferenceElement` as, which is of type: `Element | VirtualElement` where `VirtualElement` is defined to this type: `{getBoundingClientRect: () => ClientRect | DOMRect, contextElement?: Element};`. But I'm not sure if we want to do this, or if it is just easier to keep it as it is now with the type assertion `as ReferenceElement`. Wdyt? :)

Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: However after thinking on it, maybe it's better if targetEl is of type `ReferenceElement | null` in the `AttentionState` type. Then it would be enough to have this: `const referenceEl = state?.targetEl`. Then I would just have to also change the type of `targetEl` in our react-repo to also be of type: `ReferenceElement | null`. However, I would have to create my own `ReferenceElement` type in the react-repo since that type is defined by `floating-ui/dom` and I don't think we want to add `floating-ui/dom` as a devDependency to just to be able to use their `ReferenceElement` type. Wdyt? :)

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: I tried that! But then I also need to set the direction enums to of type `Directions`: 
```
const TOP_START: Directions = 'top-start'
const TOP: Directions = 'top'
const TOP_END: Directions = 'top-end'
const RIGHT_START: Directions = 'right-start'
const RIGHT: Directions = 'right'
const RIGHT_END: Directions = 'right-end'
const BOTTOM_START: Directions = 'bottom-start'
const BOTTOM: Directions = 'bottom'
const BOTTOM_END: Directions = 'bottom-end'
const LEFT_START: Directions = 'left-start'
const LEFT: Directions = 'left'
const LEFT_END: Directions = 'left-end'
```
That seems to remove the error for when I'm setting `arrowRotation` to `rotation[arrowDirection]`. However, now I instead get an error regarding the type of the `name`-parameter that is being passed into `isDirectionVertical`:
```
const isDirectionVertical = (name: Directions) =>
  [TOP_START, TOP, TOP_END, BOTTOM_START, BOTTOM, BOTTOM_END].includes(name)
``` 
I get this error then: 
```
Argument of type 'Directions' is not assignable to parameter of type '"top" | "top-start" | "top-end" | "bottom" | "bottom-start" | "bottom-end"'.
  Type '"right"' is not assignable to type '"top" | "top-start" | "top-end" | "bottom" | "bottom-start" | "bottom-end"'.
```

File: package.json
Comment (felicia-haggqvist) [reply package.json]: Updated in commit: [6c52c9d](https://github.com/warp-ds/core/pull/15/commits/6c52c9dd87f76c84b60c6c03389dbcd4302f36c2#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R36)

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: Updated in commit: [6c52c9d](https://github.com/warp-ds/core/pull/15/commits/6c52c9dd87f76c84b60c6c03389dbcd4302f36c2#diff-f7f359dac756e644d43cfb407c55dca65dc06f8c392dd99f6a13681375c1cc10R11)

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: Updated in commit: [6c52c9d](https://github.com/warp-ds/core/pull/15/commits/6c52c9dd87f76c84b60c6c03389dbcd4302f36c2#diff-f7f359dac756e644d43cfb407c55dca65dc06f8c392dd99f6a13681375c1cc10L105)

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: Updated in commit: [6c52c9d](https://github.com/warp-ds/core/pull/15/commits/6c52c9dd87f76c84b60c6c03389dbcd4302f36c2#diff-f7f359dac756e644d43cfb407c55dca65dc06f8c392dd99f6a13681375c1cc10R113)

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: Updated in commit: [6c52c9d](https://github.com/warp-ds/core/pull/15/commits/6c52c9dd87f76c84b60c6c03389dbcd4302f36c2#diff-f7f359dac756e644d43cfb407c55dca65dc06f8c392dd99f6a13681375c1cc10R143)

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: Updated in commit: [6c52c9d](https://github.com/warp-ds/core/pull/15/commits/6c52c9dd87f76c84b60c6c03389dbcd4302f36c2#diff-f7f359dac756e644d43cfb407c55dca65dc06f8c392dd99f6a13681375c1cc10R141)

File: src/attention/utils/helpers.ts
Comment (BalbinaK) [inline src/attention/utils/helpers.ts]: Would it work if you reused `opposites[side]` here?

Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: It should work. Just need to figure out how to solve the type issues. I get some type errors when trying to use `opposites[side]`.

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: Updated in this commit: [2a7276c](https://github.com/warp-ds/core/pull/15/commits/2a7276c3784d4a7cb8e14b7a1a790a167781024a#diff-f7f359dac756e644d43cfb407c55dca65dc06f8c392dd99f6a13681375c1cc10R115)

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: Updated in this commit: [2a7276c](https://github.com/warp-ds/core/pull/15/commits/2a7276c3784d4a7cb8e14b7a1a790a167781024a#diff-f7f359dac756e644d43cfb407c55dca65dc06f8c392dd99f6a13681375c1cc10R168)

File: package.json
Comment (imprashast) [inline package.json:35]: Since we are not building `core`, this dependency won't end up in the client's package lock unless specified here or in `peerDependencies`

File: src/attention/utils/helpers.ts
Comment (imprashast) [inline src/attention/utils/helpers.ts]: Maybe refer to this from the const above that's a type of `Directions`. This currently looks like a string without checking the type

File: src/attention/utils/helpers.ts
Comment (imprashast) [inline src/attention/utils/helpers.ts]: Maybe safeguard `middlewareData?.arrow`?

**Review by imprashast — APPROVED** [review]
Few suggestions, look good apart from those. Good job 👍

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: But isn't it already set as type `Directions` here? We are setting `actualDirection`, `directionName` and `arrowEl` to be part of the `AttentionState` type. When I hover on `directionName` it says that it is already of type `Directions`:

Skärmavbild 2024-03-04 kl  09 57 00

 and if I try and replace `bottom` with just a random string, it complains saying that it is not part of the `Directions` type: 
Skärmavbild 2024-03-04 kl  09 57 09

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: Updated in this commit: [a7fe061](https://github.com/warp-ds/core/pull/15/commits/a7fe061a8bd466af2a2fa382079595ea988beb9d)

File: package.json
Comment (felicia-haggqvist) [reply package.json:35]: Updated in this commit: [157a05f](https://github.com/warp-ds/core/pull/15/commits/157a05f70416376c62d76fee4204ce06a72c01b5)

File: src/attention/utils/helpers.ts
Comment (imprashast) [reply src/attention/utils/helpers.ts]: Yes it does but without hovering, it looks like a string. Since we already have a typed const in the file, wouldn't it be better to use that instead of re-defining a string?

Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: Which typed const in the file are you referring to that I should use instead of re-defining a string? 🙈

Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: I could create a `defaultDirectionName` const and use that instead:   `const defaultDirectionName: Directions = 'bottom'`? Like this: [992972f](https://github.com/warp-ds/core/pull/15/commits/992972fa4d611cab12c3ac9e62d0645e36009a34)

Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: Like this: [992972f](https://github.com/warp-ds/core/pull/15/commits/992972fa4d611cab12c3ac9e62d0645e36009a34)

Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: Updated to use `BOTTOM` enum instead of `defaultDirectionName` in this commit: [5056e3b](https://github.com/warp-ds/core/pull/15/commits/5056e3b31087983ad8bc9ff71b4259d5bbad0838)

**Review by imprashast — APPROVED** [review]

**Comment by BalbinaK** [comment]
Not a big deal, but when I was testing React Attention with core and css changes locally, I saw that the arrow position is slightly too low (at least in the case of `highlight`):
Screenshot 2024-03-07 at 16 17 06

**Comment by felicia-haggqvist** [comment]
@BalbinaK Ah nice catch! I’ll look into it. 😄 👍

File: package.json
Comment (BalbinaK) [inline package.json:28]: If @floating-ui/dom is a peerDependency, do we still need to have it under devDependencies?

File: src/attention/utils/helpers.ts
Comment (BalbinaK) [inline src/attention/utils/helpers.ts]: Can it just be `return direction`?

**Review by BalbinaK — COMMENTED** [review]
Should Attention be tested on a server-side-rendered test app before we merge this PR?

File: package.json
Comment (felicia-haggqvist) [reply package.json:28]: Not sure! @imprashast Do you know if we should have `@floating-ui/dom` both as a peerDependency and a devDependency?

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: You're right! 👍 I must have forgotten to change it when I refactored it before! 🙈

File: package.json
Comment (imprashast) [reply package.json:28]: Since we are not building anything in this app, we don't need it under devDep but just peerDep :)

File: src/attention/utils/helpers.ts
Comment (felicia-haggqvist) [reply src/attention/utils/helpers.ts]: Updated in this commit: [5440729](https://github.com/warp-ds/core/pull/15/commits/54407296b1cb66809b04d2f2312b47026eff4ff8)

File: package.json
Comment (felicia-haggqvist) [reply package.json:28]: Updated in this commit: [ea3c6c1](https://github.com/warp-ds/core/pull/15/commits/ea3c6c142f3a98523b57389bf5bb8bb2d76aa3b3)

**Review by BalbinaK — APPROVED** [review]
LGTM provided it also works SSR 😄

**Comment by felicia-haggqvist** [comment]
@BalbinaK Tested vue, react and elements in a server-side-rendered test app and it looks good! 🎉

**Pull request merged**
